core/src/worker/tuner/resource_based.rs

```diff
@@ -522,25 +520,74 @@ impl RealSysInfo {
             self.cur_mem_usage.store(mem, Ordering::Release);
         }
         self.cur_cpu_usage.store(cpu.to_bits(), Ordering::Release);
-        self.last_refresh.store(Instant::now());
+    }
+}
+
+/// Tracks host resource usage by refreshing metrics on a background thread.
+pub struct RealSysInfo {
+    inner: Arc<RealSysInfoInner>,
+    shutdown: Arc<AtomicBool>,
+    shutdown_handle: Mutex<Option<thread::JoinHandle<()>>>,
+}
+
+impl RealSysInfo {
+    pub(crate) fn new() -> Self {
+        let mut sys = sysinfo::System::new();
+        sys.refresh_memory();
+        let total_mem = sys.total_memory();
+        let inner = Arc::new(RealSysInfoInner {
+            sys: Mutex::new(sys),
+            cur_mem_usage: AtomicU64::new(0),
+            cur_cpu_usage: AtomicU64::new(0),
+            total_mem: AtomicU64::new(total_mem),
+        });
+        inner.refresh();
+
+        let shutdown = Arc::new(AtomicBool::new(false));
+        let thread_inner = inner.clone();
+        let thread_shutdown = shutdown.clone();
+        let handle = thread::Builder::new()
+            .name("temporal-real-sysinfo".to_string())
+            .spawn(move || {
+                const REFRESH_INTERVAL: Duration = Duration::from_millis(100);
+                loop {
+                    if thread_shutdown.load(Ordering::Acquire) {
+                        return;
+                    }
+                    thread_inner.refresh();
+                    thread::sleep(REFRESH_INTERVAL);
+                }
+            })
+            .expect("failed to spawn RealSysInfo refresh thread");
+
+        Self {
+            inner,
+            shutdown,
+            shutdown_handle: Mutex::new(Some(handle)),
+        }
     }
 }
 
 impl SystemResourceInfo for RealSysInfo {
     fn total_mem(&self) -> u64 {
-        self.total_mem.load(Ordering::Acquire)
+        self.inner.total_mem.load(Ordering::Acquire)
     }
 
     fn used_mem(&self) -> u64 {
-        // TODO: This should really happen on a background thread since it's getting called from
-        //   the async reserve
-        self.refresh_if_needed();
-        self.cur_mem_usage.load(Ordering::Acquire)
+        self.inner.cur_mem_usage.load(Ordering::Acquire)
     }
 
     fn used_cpu_percent(&self) -> f64 {
-        self.refresh_if_needed();
-        f64::from_bits(self.cur_cpu_usage.load(Ordering::Acquire))
+        f64::from_bits(self.inner.cur_cpu_usage.load(Ordering::Acquire))
+    }
+}
+
+impl Drop for RealSysInfo {
+    fn drop(&mut self) {
+        self.shutdown.store(true, Ordering::Release);
+        if let Some(handle) = self.shutdown_handle.lock().take() {
+            let _ = handle.join();
+        }
```

Comment:
In the worst case this blocks for 100ms which isn't a huge deal, but, not the nicest behavior. You can interrupt your sleep by using https://doc.rust-lang.org/std/thread/fn.park_timeout.html or a channel with https://doc.rust-lang.org/1.54.0/std/sync/mpsc/struct.Receiver.html#method.recv_timeout